mesonbuild/dependencies/dub.py

```diff
@@ -300,12 +309,106 @@ def find_package_target(pkg: T.Dict[str, str]) -> bool:
+        # We do show a WARNING that the build type is not the same.
+        # It might be critical in release builds, and acceptable otherwise
+        check_list = {'configuration', 'platform', 'arch', 'compiler', 'compiler_version'}
+
+        mlog.debug('Searching in DUB cache for compatible', spec['target_file_name'])
+
+        if self.use_cache_db:
+            result = self._find_in_cache_db(spec, check_list)
+        else:
+            result = self._find_in_old_cache(spec, check_list)
+
+        if result[0] is not None:
+            mlog.debug('Found', result[0])
+
+        return result
+
+    def _find_in_cache_db(self, spec: JsonDict, check_list: T.Set[str]) -> T.Tuple[str, T.Set[str]]:
+        # The cache database is a JSON file written by newer versions of Dub to help to locate builds
+        # artifact in the cache.
+        # The JSON file path is ~/.dub/cache/pkg/version/[+subpkg/]db.json
+
+        if not DubDependency.class_cache_dir:
+            DubDependency.class_cache_dir = self._find_cache_dir()
+        cache_dir = DubDependency.class_cache_dir
+
+        pkg = spec['name']
+        subpkg = None
+        if ':' in pkg:
+            [pkg, subpkg] = pkg.split(':')
+        pkg_cache_dir = os.path.join(cache_dir, pkg, spec['version'])
+        if subpkg is not None:
+            pkg_cache_dir = os.path.join(pkg_cache_dir, f'+{subpkg}')
+        db_file = os.path.join(pkg_cache_dir, 'db.json')
+
+        if not os.path.exists(db_file):
+            return (None, None)
+
+        mlog.debug('Checking in DUB cache database', db_file)
+
+        compatibilities: T.Set[str] = set()
+
+        with open(db_file) as f:
+            db = json.load(f)
+            for entry in db:
+                target = entry['targetBinaryPath']
+                if not os.path.exists(target):
+                    mlog.debug("WARNING: Could not find a Dub target: " + target)
+                    continue
+
+                comps = set()
+                if spec['configuration'] == entry['configuration']:
+                    comps.add('configuration')
+
+                if spec['build_type'] == entry['buildType']:
+                    comps.add('build_type')
+
+                if all(platform in entry['platform'] for platform in spec['platform']):
+                    comps.add('platform')
+
+                if all(arch in entry['architecture'] for arch in spec['arch']):
+                    comps.add('arch')
+
+                if spec['compiler'] == entry['compiler']:
+                    comps.add('compiler')
+
+                if not spec['compiler_version'] or any(cv == entry['compiler_version'] for cv in spec['compiler_version']):
+                    comps.add('compiler_version')
+
+                if all(key in comps for key in check_list):
+                    return (target, comps)
+                else:
+                    compatibilities = set.union(compatibilities, comps)
+
+        return (None, compatibilities)
```

Comment:
All this code is duplicated between functions, it seems? I think what you should do is hoist that out so it doesn't need to be duplicated...